tests/test_typing.py

```diff
@@ -0,0 +1,158 @@
+        param(SubscriptedList, "list", [0, 1, None], True, id="list_elt_partial"),
+        param(SubscriptedDict, "dict_str", {"key": None}, True, id="dict_elt"),
+        param(
+            SubscriptedDict,
+            "dict_str",
+            {"key_valid": 123, "key_invalid": None},
+            True,
+            id="dict_elt_partial",
+        ),
+        param(SubscriptedList, "list", None, True, id="list"),
+        param(SubscriptedDict, "dict_str", None, True, id="dict"),
+        param(SubscriptedList, "opt_list", [None], True, id="opt_list_elt"),
+        param(SubscriptedDict, "opt_dict_str", {"key": None}, True, id="opt_dict_elt"),
+        param(SubscriptedList, "opt_list", None, False, id="opt_list"),
+        param(SubscriptedDict, "opt_dict_str", None, False, id="opt_dict"),
+        param(SubscriptedList, "list_opt", [None], False, id="list_opt_elt"),
+        param(SubscriptedDict, "dict_opt", {"key": None}, False, id="dict_opt_elt"),
+        param(SubscriptedList, "list_opt", None, True, id="list_opt"),
+        param(SubscriptedDict, "dict_opt", None, True, id="dict_opt"),
+        param(
+            ListConfig([None], element_type=Optional[User]),
+            0,
+            User("Bond", 7),
+            False,
+            id="Optional[User]<-User_succeeds",
+        ),
+        param(
+            ListConfig([User], element_type=User),
+            0,
+            None,
+            True,
+            id="Optional[User]<-str_fails",
+        ),
+    ],
+)
+def test_assign(cls: Any, key: str, assignment: Any, error: bool) -> None:
+    cfg = OmegaConf.structured(cls)
+    if error:
+        with raises(ValidationError):
+            cfg[key] = assignment
+    else:
+        cfg[key] = assignment
+        assert cfg[key] == assignment
+
+
+@mark.parametrize(
+    "src,op,keys,ref_type,is_optional",
+    [
+        param(Group, None, ["admin"], User, True, id="opt_user"),
+        param(
+            ConcretePlugin,
+            None,
+            ["params"],
+            ConcretePlugin.FoobarParams,
+            False,
+            id="nested_structured_conf",
+        ),
+        param(
+            OmegaConf.structured(Users({"user007": User("Bond", 7)})).name2user,
+            None,
+            ["user007"],
+            User,
+            False,
+            id="structured_dict_of_user",
+        ),
+        param(
+            DictConfig({"a": 123}, element_type=int),
+            None,
+            ["a"],
+            int,
+            False,
+            id="dict_int",
+        ),
+        param(
+            DictConfig({"a": 123}, element_type=Optional[int]),
+            None,
+            ["a"],
+            int,
+            True,
+            id="dict_opt_int",
+        ),
+        param(DictConfig({"a": 123}), None, ["a"], Any, True, id="dict_any"),
+        param(
+            ListConfig([], element_type=int),
+            lambda cfg: cfg.insert(0, 123),
```

Comment:
I think you could test the same thing without introducing an extra `op` with:
```python
ListConfig([], element_type=int) + [123]
```

Reply:
I was trying to test `ListConfig.insert` specifically, since the `insert` method required an update to handle `element_type` properly.
In the case of `ListConfig([], element_type=int) + [123]`, The `ListConfig.append` method is used instead of `ListConfig.insert`.

This being said, I'm becoming suspicious about the differences between `ListConfig.insert` and `ListConfig.append`. Why should `insert` need to access `self._metadata.element_type` when `append` does not? Perhaps something is missing from the logic in `append`, or maybe the logic in `insert` could be simplified.

Reply:
Ah, I think I had read the code too quickly and misunderstood where the related change was taking place, never mind.

I think `append()` should check the `element_type` through `_set_item_impl()` (in `wrap()`). But I also think that `insert()` could be simplified in a way similar to `append()`, by first inserting `None` into `_content` then setting its value.

Reply:
I'm able to get the tests to pass after removing the `_maybe_wrap()` call from `insert()`, and calling `_set_item_impl()` directly like in `append()`.

Reply:
> I'm able to get the tests to pass after removing the `_maybe_wrap()` call from `insert()`, and calling `_set_item_impl()` directly like in `append()`.

Cool, maybe you can leave it for another PR to avoid adding even more stuff into that one.

Reply:
Regarding the addition of `op` that is used in only one test case, you can use a trick like this:
```python
(lambda cfg: cfg.insert(0, 123) or cfg)(ListConfig([], element_type=int))
```
Though if you go this route it should definitely be commented to explain what's going on here... not sure if it's really a better option, but I thought I'd mention it anyway :)

Reply:
It looks like you made the 15-20 test cases + the test code itself more complicated for the benefit of a single test case.
In such cases, I prefer to create a dedicated test that handles the special case.
I would also rename test_ref_type to test_something, e.g. test_assignment_optional_behavior and test_list_insert_optional_behavior. (also, what about testing of ListConfig.append?)

Reply:
> It looks like you made the 15-20 test cases + the test code itself more complicated for the benefit of a single test case.
> In such cases, I prefer to create a dedicated test that handles the special case.

Got it. Done in https://github.com/omry/omegaconf/pull/749/commits/c488af67e6aa3ac2f8ccb0b81b3812f2039f4887.

> I would also rename test_ref_type to test_something, e.g. test_assignment_optional_behavior and test_list_insert_optional_behavior.

Renamed in https://github.com/omry/omegaconf/pull/749/commits/aeda10ccb86cf0a45c03ab0627963ff215d43409.

> (also, what about testing of ListConfig.append?)

I've added some tests for append in https://github.com/omry/omegaconf/pull/749/commits/ceb5dd1922f0582b315a7d067e265eec90cc7d80.
Originally I tested `insert` only because this PR does not touch `append`'s code.

After some discussion (see the above), I became suspicious about why `append`'s code is so different from `insert`'s -- it seems that the two have diverged. I'm planning a followup PR to reconcile this.